packages/discussions/lib/discussionPreferences.js

```diff
@@ -80,6 +81,92 @@ module.exports = async ({
   if (existingDP) {
     await transaction.public.discussionPreferences.updateOne(findQuery, updateQuery, options)
   } else {
-    await transaction.public.discussionPreferences.insert(updateQuery, options)
+    await Promise.all([
+      transaction.public.discussionPreferences.insert(updateQuery, options),
+      loaders.Discussion.Commenter.discussionPreferences.clear({
+        userId: userId,
+        discussionId: discussion.id
+      })
+    ])
+  }
+
+  await ensureAnonymousDifferentiator({
+    transaction,
+    userId,
+    discussion,
+    t,
+    loaders
+  })
+}
+
+const ensureAnonymousDifferentiator = async ({
+  transaction,
+  userId,
+  discussion,
+  t,
+  loaders
+}) => {
+  // const discussionId = discussion.id
+  const discussionId = discussion.id
+  const findQuery = {
+    userId, discussionId
+  }
+
+  const preferences = await transaction.public.discussionPreferences.findOne(findQuery)
+
+  if (preferences && preferences.anonymousDifferentiator) {
+    return
+  }
+
+  const userHasComments = !!await transaction.public.comments.findFirst({
+    userId, discussionId
+  })
```

Comment:
```suggestion
  const userHasComments = await transaction.public.comments.count({
    userId, discussionId
  }) > 0
```

Reply:
Is this a matter of style?

I feel, that as someone who reads the code, it is more likely to miss the `> 0` at the end of a line than the `!!` at the beginning of a line.

Anyway, not agreing on how to write the type cast, tells me that I should go for this:

```js
const amountOfComments = await transaction.public.comments.count({
   userId, discussionId
})

if (amountOfComments === 0) {
  return
}
```

Reply:
I'm also fine with
```suggestion
  const userHasComments = !!await transaction.public.comments.count({
    userId, discussionId
  })
```

My main suggestion is using `count` instead of `findFirst` since we don't need the data.